state/store.go

```diff
@@ -366,6 +389,51 @@ func (store dbStore) PruneStates(from int64, to int64, evidenceThresholdHeight i
 	return nil
 }
 
+// PruneABCIResponses attempts to prune all ABCI responses up to, but not
+// including, the given height. On success, returns the number of heights
+// pruned and the new retain height.
+func (store dbStore) PruneABCIResponses(targetRetainHeight int64) (int64, int64, error) {
+	if store.DiscardABCIResponses {
+		return 0, 0, errors.New("ABCI responses are discarded, nothing to prune")
+	}
+	lastRetainHeight, err := store.getLastABCIResponsesRetainHeight()
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed to look up last ABCI responses retain height: %w", err)
+	}
+	if lastRetainHeight == 0 {
+		lastRetainHeight = 1
+	}
+
+	batch := store.db.NewBatch()
+	defer batch.Close()
+
+	pruned := int64(0)
+	batchPruned := int64(0)
+
+	for h := lastRetainHeight; h < targetRetainHeight; h++ {
+		if err := batch.Delete(calcABCIResponsesKey(h)); err != nil {
+			return pruned, lastRetainHeight + pruned, fmt.Errorf("failed to delete ABCI responses at height %d: %w", h, err)
+		}
+		batchPruned++
+		if batchPruned%1000 == 0 && batchPruned > 0 {
```

Comment:
Hmm, probably nothing, but it would not make sense to use `>=`, `==` would have the same effect as `%`

Reply:
`>='  or `==` are both good to me (I tend to use the former as it is more "defensive")

The point of my question was that the current if is not very "readable"... it makes you think that if will periodically be activated, whereas it is not the case (BTW, this is an "idiomatic" comment 😄)

Reply:
Yes, you're right. The choice of `>=` (defensive) is in case there's a bug we didn't catch inside the loop that would (e.g, skip the value where `==` becomes true, effectively turning it into an infinite loop).
But this is super-minor, you can keep the `==`, no problem with that.

Reply:
Hah that is a new level of defensive :D Thanks for sharing that, no worries, I can change it to `>=` :D